librz/type/path.c

```diff
@@ -122,91 +122,140 @@ RZ_API st64 rz_type_offset_by_path(const RzTypeDB *typedb, RZ_NONNULL const char
+				if (tpath) {
+					rz_list_append(list, tpath);
+				}
+			}
+			collect_type_paths(typedb, list, memb->type, newpath, offset, depth - 1);
+			if (offset != 0) {
+				free(newpath);
+			}
+		}
+		break;
+	}
+	case RZ_BASE_TYPE_KIND_TYPEDEF: {
+		RzType *ttype = rz_type_db_base_type_unwrap_typedef(typedb, btype);
+		if (!ttype) {
+			return;
 		}
-	} else if (type->identifier.kind == RZ_TYPE_IDENTIFIER_KIND_UNION) {
-		// Get the base type
+		collect_type_paths(typedb, list, ttype, prefix, offset, depth); // depth not decreased because this is not a visible step in the path
+		break;
+	}
+	default:
+		break;
+	}
+}
+
+static void collect_type_paths(const RzTypeDB *typedb, RzList /*<RzTypePath *>*/ *list, RzType *type, char *prefix, ut64 offset, unsigned int depth) {
+	rz_return_if_fail(typedb && list && type && prefix);
+	if (!depth) {
+		return;
+	}
+	switch (type->kind) {
+	case RZ_TYPE_KIND_IDENTIFIER: {
+		rz_return_if_fail(type->identifier.name);
 		RzBaseType *btype = rz_type_db_get_base_type(typedb, type->identifier.name);
 		if (!btype) {
-			return false;
+			return;
 		}
-		RzTypeUnionMember *memb;
-		rz_vector_foreach(&btype->union_data.members, memb) {
-			char *newpath = rz_str_newf("%s.%s", path, memb->name);
-			result &= structured_member_walker(typedb, list, parent, memb->type, path, offset);
+		collect_base_type_paths(typedb, list, btype, prefix, offset, depth); // depth not decreased because this is not a visible step in the path
+		break;
+	}
+	case RZ_TYPE_KIND_ARRAY: {
+		rz_return_if_fail(type->array.type);
+		ut64 stride = rz_type_db_get_bitsize(typedb, type->array.type) / 8;
+		if (!stride) {
+			break;
+		}
+		ut64 idx = offset / stride;
+		if (idx >= type->array.count) {
+			break;
+		}
+		ut64 idx_offset = offset - idx * stride;
+		char *newpath = rz_str_newf("%s[%" PFMT64u "]", prefix, idx);
+		if (idx_offset == 0) {
+			RzTypePath *tpath = rz_type_path_new(type->array.type, newpath);
+			if (tpath) {
+				rz_list_append(list, tpath);
+			}
+		}
+		collect_type_paths(typedb, list, type->array.type, newpath, idx_offset, depth - 1);
+		if (idx_offset != 0) {
 			free(newpath);
 		}
+		break;
+	}
+	default:
+		break;
 	}
-	return result;
 }
 
 /**
- * \brief Returns the list of all type paths matching the offset
+ * \brief Returns the list of all paths into the base type matching the offset
  *
- * \param typedb Types Database instance
- * \param btype The base type
- * \param offset The offset of the path to match against
+ * \param btype The root type from which paths are searched
+ * \param offset The offset into \p btype to match against
+ * \param max_depth Maximum number of components a path may have
  */
-RZ_API RZ_OWN RzList /*<RzTypePath *>*/ *rz_type_path_by_offset(const RzTypeDB *typedb, RzBaseType *btype, ut64 offset) {
+RZ_API RZ_OWN RzList /*<RzTypePath *>*/ *rz_base_type_path_by_offset(const RzTypeDB *typedb, RzBaseType *btype, ut64 offset, unsigned int max_depth) {
```

Comment:
Maybe worth writing that `RzBaseType *btype` is also `const`?

Reply:
Good point